crates/router_env/src/request_id.rs

```diff
@@ -0,0 +1,365 @@
+#[derive(Clone, Debug)]
+pub struct RequestIdentifier {
+    header_name: String,
+    use_incoming_id: IdReuse,
+}
+
+/// Request ID value that can be extracted in route handlers.
+///
+/// Wraps an `Arc<str>` for optimal performance in web middleware.
+#[derive(Clone, PartialEq, Eq, Debug, Hash)]
+pub struct RequestId(Arc<str>);
+
+/// The middleware implementation that processes requests.
+#[derive(Debug)]
+pub struct RequestIdMiddleware<S> {
+    service: S,
+    header_name: HeaderName,
+    use_incoming_id: IdReuse,
+}
+
+impl Display for RequestId {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.as_str())
+    }
+}
+
+impl FromStr for RequestId {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.is_empty() {
+            Err(Error::InvalidHeaderValue {
+                value: s.to_string(),
+            })
+        } else {
+            Ok(Self(s.into()))
+        }
+    }
+}
+
+impl TryFrom<HeaderValue> for RequestId {
+    type Error = Error;
+
+    fn try_from(value: HeaderValue) -> Result<Self, Self::Error> {
+        let s = value.to_str().map_err(|_| Error::InvalidHeaderValue {
+            value: format!("{:?}", value),
+        })?;
+        Self::from_str(s)
+    }
+}
+
+impl From<RequestId> for HeaderValue {
+    fn from(request_id: RequestId) -> Self {
+        Self::from_str(&request_id.0).unwrap_or_else(|_| Self::from_static("invalid-request-id"))
+    }
+}
+
+impl From<RequestId> for String {
+    fn from(request_id: RequestId) -> Self {
+        request_id.0.to_string()
+    }
+}
+
+impl From<String> for RequestId {
+    fn from(s: String) -> Self {
+        Self(s.into())
+    }
+}
+
+impl From<&str> for RequestId {
+    fn from(s: &str) -> Self {
+        Self(s.into())
+    }
+}
+
+impl RequestId {
+    /// Create a new RequestId from a string.
+    pub fn new_from_string(value: impl Into<Arc<str>>) -> Self {
+        Self(value.into())
+    }
+
+    /// Extract request ID from ServiceRequest header or generate UUID v7.
+    ///
+    /// This is the core logic: try to extract from the specified header,
+    /// if not possible or not desired, generate a new UUID v7.
+    pub fn extract_or_generate(
+        request: &ServiceRequest,
+        header_name: &HeaderName,
+        use_incoming_id: IdReuse,
+    ) -> Self {
+        let request_id_string = match use_incoming_id {
+            IdReuse::UseIncoming => {
+                // Try to extract from incoming header
+                if let Some(existing_header) = request.headers().get(header_name) {
+                    existing_header
+                        .to_str()
+                        .map(|s| s.to_string())
+                        .unwrap_or_else(|_| {
+                            tracing::warn!("Invalid request ID header, generating new UUID v7");
+                            generate_uuid_v7()
+                        })
```

Comment:
Can't we take advantage of the `TryFrom<HeaderValue>` implementation we added above?


Something like:
```suggestion
                    Self::try_from(existing_header)
                        .unwrap_or_else(|_| {
                            tracing::warn!("Invalid request ID header, generating new UUID v7");
                            Self(generate_uuid_v7())
                        })
```

Additionally, we can consider logging the error.